### ISSUE TITLE
"error: Incompatible return value type" when returning List[subclass of T] instead of List[T]

### ISSUE THREAD

**Issue by jstasiak** [python/mypy#2018]
CPython 3.5.2, mypy 2e2a94e30724f061a835eb2f72a89115250a37ee

Test code:

```python
from typing import List

class A:
    pass


class AA(A):
    pass


def get() -> List[A]:
    a = [AA()]  # type: List[AA]
    return a
```

Result:

```
stuff.py: note: In function "get":
stuff.py:13: error: Incompatible return value type (got List[AA], expected List[A])
```

What's interesting is that if I replace the above ``get`` definition with the following one mypy is ok with it:

```python
def get() -> List[A]:
    return [AA()]
```

**Comment by ddfisher** [comment]
I know this seems odd, but this is actually expected behavior because `List` is invariant over its contents.  Mypy will infer a wider type for a list literal from context, but it's not valid in general to do that for a List-typed variable.

Note that writing `get` like this is also valid:
```py
def get() -> List[A]:
    a = [AA()]  # type: List[A]
    return a
```
Or like this (because `Iterable` is covariant):
```py
def get() -> Iterable[A]:
    a = [AA()]  # type: List[AA]
    return a
```

**Comment by aldanor** [comment]
Hmm... something I don't understand, if `List` is invariant over the contents, why doesn't this raise any errors? (3.5.2, mypy both 0.4.3 and latest master)

```python
from typing import List

class A: pass
class B(A): pass
class C: pass

def f(a: List[A]) -> List[A]:
    return [A(), B()]  # works?
    # return [A(), B(), C()]  # List item 2 has incompatible type "C"

f([A(), B()])  # works?
# f(A(), B(), C())  # List item 2 has incompatible type "C"
```

**Comment by gvanrossum** [comment]
There's some additional subtlety in mypy's type inference. When you write
```
a = [B(), B()]
f(a)
```
the type of `[B(), B()]` (and hence of `a`) is inferred as `List[B]`, which is incompatible with `List[A]` because of the variance rules. But when you write
```
f([B(), B()])
```
the type required for f's argument, i.e. `List[A]`, is used as a *context* for deciding what the type of `[B(), B()]` should be, and then it is decided that since each individual item is compatible with `A`, the type of the list can be `List[A]`, and that's what's inferred.

So context is king!

There's a bit about this in the docs at http://mypy.readthedocs.io/en/latest/type_inference_and_annotations.html but maybe we need to add more examples?

**Comment by aldanor** [comment]
Ah, I see, thanks. I've realized there's a whole bunch of issues in mypy repo related to confusion around `List` type parameter covariance, and how the type may be deduced differently given a context. Maybe it would help indeed to have this kind of thing highlighted in the docs with a big red exclamation sign!

**Comment by gvanrossum** [comment]
We definitely need better docs for this. I notice that this section lists
covariance as not yet documented:
http://mypy.readthedocs.io/en/latest/additional_features.html

**Comment by ddfisher** [comment]
Opened #2034 to track the docs issue.
